### PR TITLE
Fix equality check when arrays are not contiguous

### PR DESCRIPTION
**Pull request by Luthaf**
This was breaking Labels equality in torch down the line.

xref https://github.com/metatensor/metatrain/pull/1128

# Contributor (creator of pull-request) checklist

 - [x] Tests updated (for new features and bugfixes)?
 - [ ] ~Documentation updated (for new features)?~
 - [ ] ~Issue referenced (for PRs that solve an issue)?~

# Reviewer checklist

 - [ ] ~CHANGELOG updated with public API or any other important changes?~

### CODE REVIEW

File: metatensor-core/include/metatensor/arrays.hpp [hunk 1]
```diff
@@ -158,21 +191,66 @@ namespace details {
...
+        auto indices = std::vector<int64_t>(ndim, 0);
+        const auto* lhs_ptr = lhs.data();
+        const auto* rhs_ptr = rhs.data();
+
+        for (size_t i = 0; i < n_elements; i++) {
+            int64_t lhs_offset = 0;
+            int64_t rhs_offset = 0;
+            for (size_t dim = 0; dim < ndim; dim++) {
+                lhs_offset += indices[dim] * lhs_strides[dim];
+                rhs_offset += indices[dim] * rhs_strides[dim];
```
Comment (HaoZeke) [inline metatensor-core/include/metatensor/arrays.hpp:236]: This seems inefficient? It's linear offset for every element?

Maybe a private tracking tuple for the offsets and an incremental update would work better but .. can also leave it for later.

File: metatensor-core/include/metatensor/arrays.hpp [hunk 2]
```diff
@@ -604,15 +694,14 @@ class NDArray {
...
-    /// is 0.
-    bool is_empty() const {
-        for (auto s: shape_) {
-            if (s == 0) {
-                return true;
-            }
-        }
-        return false;
+    /// Get the strides of this array in number of elements.
+    std::vector<int64_t> strides() const {
```
Comment (HaoZeke) [inline metatensor-core/include/metatensor/arrays.hpp]: So this is a by-value return, so a new one each time..

File: metatensor-core/include/metatensor/arrays.hpp [hunk 3]
```diff
@@ -1619,6 +1709,16 @@ class SimpleDataArray : public metatensor::DataArrayBase {
         return data_->data();
     }
 
+    /// Get the strides of this array in number of elements.
+    std::vector<int64_t> strides() const {
+        return details::contiguous_strides(shape_);
+    }
```
Comment (HaoZeke) [inline metatensor-core/include/metatensor/arrays.hpp]: Also by-value.

File: metatensor-core/include/metatensor/arrays.hpp [hunk 4]
```diff
@@ -416,15 +503,16 @@ class DLPackArray {
...
-        for (auto s: shape_) {
-            if (s == 0) {
-                return true;
-            }
-        }
-        return false;
+    /// Get the strides of this array in number of elements.
+    const std::vector<int64_t>& strides() const & {
+        return strides_;
+    }
```
Comment (HaoZeke) [inline metatensor-core/include/metatensor/arrays.hpp:509]: But this one is by reference. Maybe get the others to also be by-reference with a private member?

File: python/metatensor_torch/tests/labels.py [hunk 5]
```diff
@@ -672,3 +672,15 @@ def forward(self, x: LabelsEntryWrap) -> LabelsEntryWrap:
...
+
+def test_equal_non_contiguous():
+    values_non_contiguous = torch.tensor([[1, 2, 3], [4, 3, 1]]).T
+    values_contiguous = values_non_contiguous.contiguous()
+
+    labels_non_contiguous = Labels(names=["a", "b"], values=values_non_contiguous)
+    labels_contiguous = Labels(names=["a", "b"], values=values_contiguous)
+
+    print("labels_non_contiguous", labels_non_contiguous)
+    print("labels_contiguous", labels_contiguous)
```
Comment (HaoZeke) [inline python/metatensor_torch/tests/labels.py]: Do we not have a hygine test preventing print statements in tests?

**Review by HaoZeke — CHANGES_REQUESTED** [review]
Some minor comments.

File: python/metatensor_torch/tests/labels.py (hunk 5 above)
Comment (Luthaf) [reply python/metatensor_torch/tests/labels.py]: We should, not sure why it did not catch it here, I'll check.

File: metatensor-core/include/metatensor/arrays.hpp (hunk 4 above)
Comment (Luthaf) [reply metatensor-core/include/metatensor/arrays.hpp:509]: fair enough!

File: metatensor-core/include/metatensor/arrays.hpp (hunk 1 above)
Comment (Luthaf) [reply metatensor-core/include/metatensor/arrays.hpp:236]: Why private tuple? This is only used in this function. We can always get back to it and speed it up later yes!

File: python/metatensor_torch/tests/labels.py (hunk 5 above)
Comment (Luthaf) [reply python/metatensor_torch/tests/labels.py]: ah no, this is only in metatomic. I'll open an issue for this repo as well, we can do it later

**Review by HaoZeke — APPROVED** [review]
LGTM, thanks @Luthaf

**Comment by HaoZeke** [comment]
Merge at will post-CI :)
